### PR TITLE
vtol_type: Added position feedback to VTOL backward transition

### PR DESCRIPTION
**Pull request by oravla5**
### Solved Problem
VTOL back-transition logic was open-loop on position, which resulted in usual over / undershoots w.r.t. the back-transition end-point.

### Solution
Added position feedback to the VTOL back-transition logic. The distance between the vehicle and the back-transition end The acceleration setpoint is then computed using a rectilinear kinematic model.

### Testing
A first version of this changes (not including some of the sanity checks) was tested on a Standard VTOL, see associated logs [here](https://review.px4.io/plot_app?log=7b77c77f-7614-45f1-a304-19838dfd8600)

### CODE REVIEW

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 1]
```diff
@@ -176,10 +177,45 @@ float VtolType::update_and_get_backtransition_pitch_sp()
 
 	const float track = atan2f(_local_pos->vy, _local_pos->vx);
 	const float accel_body_forward = cosf(track) * _local_pos->ax + sinf(track) * _local_pos->ay;
+	const float vel_body_forward = cosf(track) * _local_pos->vx + sinf(track) * _local_pos->vy;
 
-	// increase the target deceleration setpoint provided to the controller by 20%
-	// to make overshooting the transition waypoint less likely in the presence of tracking errors
-	const float dec_sp = _param_vt_b_dec_mss.get() * 1.2f;
+	// Add 20% factor to maximum allowed deceleration
+	const float default_dec_sp = _param_vt_b_dec_mss.get() * 1.2f;
```
Comment (RomanBapst) [inline src/modules/vtol_att_control/vtol_type.cpp:183]: @oravla5 Maybe we can leave away this magic factor and see how it goes with it.

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 2]
```diff
@@ -176,10 +177,45 @@ float VtolType::update_and_get_backtransition_pitch_sp()
...
+	// Maximum allowed deceleration setpoint as a function of the nominal deceleration setpoint
+	const float max_dec_sp = 2.f * default_dec_sp;
+
+	float dec_sp = default_dec_sp;
+
+	if (_attc->get_pos_sp_triplet()->current.valid) {
+		// Add position feedback to the deceleration setpoint calculation
+
+		// Compute backtransition end-point in local reference frame body x-direction -> dist_body_forward
+		MapProjection map_proj{_attc->get_local_pos()->ref_lat , _attc->get_local_pos()->ref_lon};
```
Comment (RomanBapst) [inline src/modules/vtol_att_control/vtol_type.cpp:193]: Maybe also check if ref_lat, ref_lon are valid

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 3]
```diff
@@ -176,10 +177,45 @@ float VtolType::update_and_get_backtransition_pitch_sp()
...
+			const float dist_body_forward = cosf(track) * pos_sp_dx + sinf(track) * pos_sp_dy;
+
+			if (fabsf(dist_body_forward) > FLT_EPSILON) {
+				// Compute deceleration setpoint
+				// Note this is deceleration (i.e. negative acceleration), and therefore the minus sign is skipped
+				dec_sp = vel_body_forward*vel_body_forward / (2.f * dist_body_forward);
+
+				// Check if the deceleration setpoint is within limits
+				if (!PX4_ISFINITE(dec_sp) || (dec_sp > max_dec_sp) || (dec_sp < 0.0f)) {
+					dec_sp = default_dec_sp;
```
Comment (RomanBapst) [inline src/modules/vtol_att_control/vtol_type.cpp:214]: @oravla5 I think instead of resetting to the default deceleration you can constrain the value to be within the limits.

**Comment by oravla5** [comment]
Thanks for the feedback @sfuhrer! I agree with your observation and indeed the logic to compute the acceleration setpoint logic, which in turn is used to define the pitch setpoint, could be moved somewhere else, i.e. the `TranstiionFlightTask`. 
However, I think this would be a different, way bigger change in the code, outside of the scope of this PR. The goal of this PR was just replacing the acceleration setpoint computation logic (from a constant value to a dynamic one). In my view, moving that logic somewhere else in the codebase would be a different task. What would you think about addressing the latter in a different PR?

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 4]
```diff
@@ -176,10 +177,50 @@ float VtolType::update_and_get_backtransition_pitch_sp()
...
+
+	if (_attc->get_pos_sp_triplet()->current.valid && _attc->get_local_pos()->xy_valid) {
+		// Add position feedback to the deceleration setpoint calculation
+
+		// Compute backtransition end-point in local reference frame body x-direction -> dist_body_forward
+		MapProjection map_proj{_attc->get_local_pos()->ref_lat, _attc->get_local_pos()->ref_lon};
+
+		float pos_sp_x, pos_sp_y = 0.f;
+		map_proj.project(_attc->get_pos_sp_triplet()->current.lat, _attc->get_pos_sp_triplet()->current.lon, pos_sp_x,
+				 pos_sp_y);
```
Comment (Jaeyoung-Lim) [inline src/modules/vtol_att_control/vtol_type.cpp:196]: Does this mean we will not be able to back transition without a valid global position estimate? Where is this being handled?

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]
@oravla5 I agree with @sfuhrer, and I am not sure if all the edge cases are being handled for this change to be able to be merged in upstream.

File: src/modules/vtol_att_control/vtol_type.cpp (hunk 4 above)
Comment (oravla5) [reply src/modules/vtol_att_control/vtol_type.cpp:196]: Hey @Jaeyoung-Lim , this is handled 7 lines above, in the `if` condition:
```
 if (_attc->get_pos_sp_triplet()->current.valid && _attc->get_local_pos()->xy_valid)
```
If the available global position is not valid, the new logic defaults to the parameter value, as before these changes :ok_hand:

**Comment by oravla5** [comment]
Hey @Jaeyoung-Lim , thanks for the feedback!
>  I am not sure if all the edge cases are being handled for this change to be able to be merged in upstream.
I understand your concerns. 
The changes I implemented check that the computed deceleration setpoint is a finite value, if not, it is set to the default constant acceleration setpoint (defined by the existing parameter). The resulting value is also guaranteed to be within 0 and the maximum deceleration setpoint value (defined as a function of the decelaration setpoint parameter). Could you be more specific about other edge cases that may not be covered here?

**Comment by Jaeyoung-Lim** [comment]
> Could you be more specific about other edge cases that may not be covered here?

For me, this is the concern. The fact that we have a cascaded control structure (att control, position control), but you are using the position as part of the logic makes it unclear how the two loops interact. I see that to some degree this is already the case, but I am not so happy to see more extensions in this direction is being made.

@oravla5 You mention the "bigger" change that is needed to change this. Why not just do it as part of this PR, so we can improve the quality of the implementation with this PR? I fear that we merge this as a quick fix, along side many more quick fixes, without any concrete plans and it will be at the same state when we arrive at the end of this year.

**Comment by RomanBapst** [comment]
@sfuhrer @oravla5 @Jaeyoung-Lim I think Silvan's input of trying to move this logic to the Transition flighttask is good.
The more logic that is related to creating attitude setpoints we can move out of this module and into the flighttask, the better. I can help @oravla5 to make the required changes (assuming for now that it's feasible).

@Jaeyoung-Lim Keep in mind that @oravla5 has just started working with PX4 a couple of months and might not always be aware of the existing "messes". So it might be more useful to make concrete suggestions, e.g. providing a link to the transition flighttask.

File: src/modules/vtol_att_control/vtol_type.cpp (hunk 4 above)
Comment (Jaeyoung-Lim) [reply src/modules/vtol_att_control/vtol_type.cpp:196]: I might be wrong, but your local position setpoint can be valid, but this doesn't mean you have a valid global position estimate. Wouldn't this introduce problems with the map projection below?

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 5]
```diff
@@ -176,10 +177,50 @@ float VtolType::update_and_get_backtransition_pitch_sp()
...
-	// increase the target deceleration setpoint provided to the controller by 20%
-	// to make overshooting the transition waypoint less likely in the presence of tracking errors
-	const float dec_sp = _param_vt_b_dec_mss.get() * 1.2f;
+	const float default_dec_sp = _param_vt_b_dec_mss.get();
+	// Maximum allowed deceleration setpoint as a function of the nominal deceleration setpoint
+	const float max_dec_sp = 2.5f * default_dec_sp;
+
+	float dec_sp = default_dec_sp;
+
+	if (_attc->get_pos_sp_triplet()->current.valid && _attc->get_local_pos()->xy_valid) {
```
Comment (Jaeyoung-Lim) [inline src/modules/vtol_att_control/vtol_type.cpp:188]: Does the position setpoint triplet become invalid if you are triggering back transition from a manual flight mode? If not, why is it relevant that the validity of the current setpoint matters?

**Comment by Jaeyoung-Lim** [comment]
> @Jaeyoung-Lim So just to be clear, your concerns are not about uncovered edge cases, but about the code structure. As I said above, I agree with these concerns.
> 
> However, adhering to atomic PR's that, when possible, address a single topic / concern is considered a good practice that eases the review process and reduces the risk of missing out things. In this case, the topic is the deceleration setpoint computation logic. This logic was already there, and these changes just try to bring an improvement over the previous implementation. The changes were tested in flight and we were able to observe a better back transition behavior of the vehicle compared to the previous logic.

I think you misunderstood my concerns. The changes in this PR, is effectively adding dependency to a position setpoint triplet, which uses global position. While dependency on the position setpoint triplet existed before, THIS pr now uses a map projection to add positional feedback to the vtol attitude control. I have added my comments regarding concerns of the edge cases with validity of global position and local position estimates in the code. However, I think the bigger concern is that because now the attitude control and position control has a circular dependency it is hard to even identify what corner cases it would have for the future. I simply don't understand the hesitancy to improve the structure within this PR, which will save you a lot of time, since it would reduce the amount of testing you would need to do to consider all the different edge cases.

I appreciate you doing the flight test, but has this been tested in cases where the back transition is triggered from manual flight modes, loss of global position estimates while the local position estimate is valid (e.g. manual flight modes) or other cases? And can we be sure that these are actually all the edge cases, and none of them have been left out? If the "clean" structure was already in place to fix this problem, these edge case testing would not have been necessary. However, if we want to make this change without improving the architecture, we need to be really sure that it doesn't break existing setups and just fix problems for the platform you are flying.

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 6]
```diff
@@ -120,3 +109,126 @@ bool FlightTaskTransition::update()
...
+{
+	return _sub_vehicle_status.get().in_transition_mode
+	       && !_sub_vehicle_status.get().in_transition_to_fw;
+}
+
+
+
+float FlightTaskTransition::computeBackTranstionPitchSetpoint()
+{
+	const vehicle_local_position_s local_pos = _sub_vehicle_local_position.get();
```
Comment (RomanBapst) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:167]:
```suggestion
	const vehicle_local_position_s &local_pos = _sub_vehicle_local_position.get();
```

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 7]
```diff
@@ -120,3 +109,126 @@ bool FlightTaskTransition::update()
...
+	const position_setpoint_s current_pos_sp = _sub_position_sp_triplet.get().current;
```
Comment (RomanBapst) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:168]:
```suggestion
	const position_setpoint_s &current_pos_sp = _sub_position_sp_triplet.get().current;
```

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.hpp [hunk 8]
```diff
@@ -66,12 +67,29 @@ class FlightTaskTransition : public FlightTask
...
+	float _decel_error_bt_int{0.}; ///< Backtransition deceleration error integrator value
+
+	const float _pitch_limit_bt = .3f; ///< Bactransition pitch limit
 
 	void updateParameters();
+	void updateParametersFromStorage();
+
+	void updateSubscribers();
+
+	bool isVtolFrontTransition();
```
Comment (RomanBapst) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.hpp:89]:
```suggestion
	bool isVtolFrontTransition() const;
```

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.hpp [hunk 9]
```diff
@@ -66,12 +67,29 @@ class FlightTaskTransition : public FlightTask
...
+	bool isVtolBackTransition();
```
Comment (RomanBapst) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.hpp:90]:
```suggestion
	bool isVtolBackTransition() const;
```

**Comment by oravla5** [comment]
@Jaeyoung-Lim @sfuhrer 
During the last week we have addressed the major concerns you shared about the initial implementation of these changes and performed some extensive testing. Our observation during testing was that the risk of overshooting during back-transition was significantly reduced. In the few cases we experienced overshooting, (situations with tail- or side-wind), its magnitude was noticeably reduced.

Find below links to some of the test flights we performed with a standard VTOL.

- [Test flight 1](https://review.px4.io/plot_app?log=1a205b3f-2faa-4c76-bd10-a9d84287383e)
- [Test flight 2](https://review.px4.io/plot_app?log=81727c60-c650-43f8-88d0-a5a9327e75b6)
- [Test flight 3](https://review.px4.io/plot_app?log=b9300f82-f1d2-42d1-9c96-e12f1dfae45c)

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 10]
```diff
@@ -120,3 +108,123 @@ bool FlightTaskTransition::update()
...
+	float deceleration_sp = default_deceleration_sp;
+
+	const float track = atan2f(local_pos.vy, local_pos.vx);
+	const float accel_body_forward = cosf(track) * local_pos.ax + sinf(track) * local_pos.ay;
+	const float vel_body_forward = cosf(track) * local_pos.vx + sinf(track) * local_pos.vy;
+
+
+	if (current_pos_sp.valid && local_pos.xy_valid) {
+
+		float pos_sp_x, pos_sp_y = 0.f;
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:181]: The calculation would look a bit more elegant if you'd use 
`Vector2f position_setpoint_local {}`;

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 11]
```diff
@@ -120,3 +108,123 @@ bool FlightTaskTransition::update()
...
+	const float max_deceleration_sp = 2.f * default_deceleration_sp;
+
+	float deceleration_sp = default_deceleration_sp;
+
+	const float track = atan2f(local_pos.vy, local_pos.vx);
+	const float accel_body_forward = cosf(track) * local_pos.ax + sinf(track) * local_pos.ay;
+	const float vel_body_forward = cosf(track) * local_pos.vx + sinf(track) * local_pos.vy;
+
+
+	if (current_pos_sp.valid && local_pos.xy_valid) {
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:179]: As you're using global position and not just local I would check if global is actually valid:
```suggestion
	if (current_pos_sp.valid && local_pos.xy_valid && local_pos.xy_global) {
```

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 12]
```diff
@@ -120,3 +108,123 @@ bool FlightTaskTransition::update()
...
+		float pos_sp_x, pos_sp_y = 0.f;
+		_geo_projection.project(current_pos_sp.lat, current_pos_sp.lon, pos_sp_x,
+					pos_sp_y);
+
+		// Compute backtransition end-point w.r.t. vehicle
+		const float pos_sp_dx = pos_sp_x - local_pos.x;
+		const float pos_sp_dy = pos_sp_y - local_pos.y;
+
+		// Compute the deceleration setpoint if the backtransition end-point is ahead of the vehicle
+		const float vel_proj = local_pos.vx * pos_sp_dx + local_pos.vy * pos_sp_dy;
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:190]: This is a simple dot product right? So how about defining all the vectors as Vector2f and then

`const bool point_ahead_of_vehicle = local_pos.dot(pos_sp_d) > FLT_EPSILON;`

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.hpp [hunk 13]
```diff
@@ -66,12 +67,29 @@ class FlightTaskTransition : public FlightTask
...
 
 	param_t _param_handle_pitch_cruise_degrees{PARAM_INVALID};
 	float _param_pitch_cruise_degrees{0.f};
+	param_t _param_handle_vt_b_dec_i{PARAM_INVALID};
+	float _param_vt_b_dec_i{0.f};
+	param_t _param_handle_vt_b_dec_mss{PARAM_INVALID};
+	float _param_vt_b_dec_mss{0.f};
 
 	AlphaFilter<float> _vel_z_filter;
+	float _decel_error_bt_int{0.}; ///< Backtransition deceleration error integrator value
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.hpp:80]:
```suggestion
	float _decel_error_bt_int{0.f}; ///< Backtransition deceleration error integrator value
```

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 14]
```diff
@@ -120,3 +108,123 @@ bool FlightTaskTransition::update()
...
+	return math::constrain(pitch_sp_new, 0.f, _pitch_limit_bt);;
+}
+
+void FlightTaskTransition::updateBackTransitioDecelerationErrorIntegrator(const float deceleration_error)
+{
+	float integrator_input = _param_vt_b_dec_i * deceleration_error;
+
+	_decel_error_bt_int += integrator_input * math::constrain(_deltatime, 0.001f, 0.1f);
+	// Saturate the integrator value
+	_decel_error_bt_int = math::constrain(_decel_error_bt_int, .0f, _pitch_limit_bt);
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:229]:
```suggestion
	_decel_error_bt_int = math::constrain(_decel_error_bt_int, 0.f, _pitch_limit_bt);
```

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 15]
```diff
@@ -120,3 +108,123 @@ bool FlightTaskTransition::update()
...
+	float deceleration_error = deceleration_sp - (-accel_body_forward);
+
+	updateBackTransitioDecelerationErrorIntegrator(deceleration_error);
+
+	return math::constrain(pitch_sp_new, 0.f, _pitch_limit_bt);;
+}
+
+void FlightTaskTransition::updateBackTransitioDecelerationErrorIntegrator(const float deceleration_error)
+{
+	float integrator_input = _param_vt_b_dec_i * deceleration_error;
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:225]:
```suggestion
	const float integrator_input = _param_vt_b_dec_i * deceleration_error;
```

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 16]
```diff
@@ -120,3 +108,123 @@ bool FlightTaskTransition::update()
...
+				}
+			}
+		}
+	}
+
+	float pitch_sp_new = _decel_error_bt_int;
+
+	// Update deceleration error integrator
+	// get accel error, positive means decelerating too slow, need to pitch up (must reverse accel_body_forward, as it is a positive number)
+	float deceleration_error = deceleration_sp - (-accel_body_forward);
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:216]:
```suggestion
	const float deceleration_error = deceleration_sp - (-accel_body_forward);
```

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 17]
```diff
@@ -120,3 +108,123 @@ bool FlightTaskTransition::update()
...
+	// Retrieve default decelaration setpoint
+	const float default_deceleration_sp = _param_vt_b_dec_mss;
+	// Maximum allowed deceleration setpoint as a function of the nominal deceleration setpoint
+	const float max_deceleration_sp = 2.f * default_deceleration_sp;
+
+	float deceleration_sp = default_deceleration_sp;
+
+	const float track = atan2f(local_pos.vy, local_pos.vx);
+	const float accel_body_forward = cosf(track) * local_pos.ax + sinf(track) * local_pos.ay;
+	const float vel_body_forward = cosf(track) * local_pos.vx + sinf(track) * local_pos.vy;
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:176]: Isn't that simply the absolute value of local_pos_xy?

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 18]
```diff
@@ -120,3 +108,123 @@ bool FlightTaskTransition::update()
...
+
+	float pitch_sp_new = _decel_error_bt_int;
+
+	// Update deceleration error integrator
+	// get accel error, positive means decelerating too slow, need to pitch up (must reverse accel_body_forward, as it is a positive number)
+	float deceleration_error = deceleration_sp - (-accel_body_forward);
+
+	updateBackTransitioDecelerationErrorIntegrator(deceleration_error);
+
+	return math::constrain(pitch_sp_new, 0.f, _pitch_limit_bt);;
```
Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:220]:
```suggestion
	return math::constrain(pitch_sp_new, 0.f, _pitch_limit_bt);
```

Comment (sfuhrer) [inline src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:220]: _decel_error_bt_int is already constraint, this additional constraint is not necessary is it?

**Review by sfuhrer — CHANGES_REQUESTED** [review]
Cool feature but I have a couple of comments about the implementation. 
- use matrix::Vector2f for geometrical calculations, then you get handy functions like .dot() and .abs() and it makes it easier to follow the code
- protect against invalid global position (local_position.xy_global)

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp [hunk 19]
```diff
@@ -100,6 +75,13 @@ bool FlightTaskTransition::activate(const trajectory_setpoint_s &last_setpoint)
 	return ret;
 }
 
+bool FlightTaskTransition::updateInitialize()
```
[the remark this answers is not in the document]
Comment (oravla5) [reply src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:78]: I used the opportunity to do some minor refactoring of the class. If definitions were moved around, it's probably because they were not following the same order as in the header file. I hope this is ok since the amount of changes is extremely minor.

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp (hunk 17 above)
Comment (oravla5) [reply src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:176]: Yeah, if I'm not mistaken, vel_body_forward would be the norm of [local_pos.vx, local_pos.vy].

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp (hunk 10 above)
Comment (oravla5) [reply src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp]: Yes, although Vector2f also consumes more flash than just two floats.

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp (hunk 18 above)
Comment (oravla5) [reply src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp]: totally right

**Comment by oravla5** [comment]
@sfuhrer I implemented the requested changes. I'm getting local build errors due to flash (roughly 700 bytes). What would be the suggested course of action in this regard. FYI @RomanBapst

File: src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/flight_mode_manager/tasks/Transition/FlightTaskTransition.cpp:78]: It's good that you take care of refactors like that, I just ask you to put it in different commits next time.
